src/main/java/io/github/thebusybiscuit/slimefun4/core/services/JsonDeserializationService.java

```diff
@@ -0,0 +1,160 @@
+     *
+     * @param clazz        The {@link Class} instance
+     * @param deserializer The deserialization logic.
+     * @param <T>          The type parameter, can be any object.
+     */
+    public <T> void registerDeserialization(@Nonnull final Class<T> clazz, @Nonnull final Function<JsonElement, ?> deserializer) {
+        deserializerMap.put(clazz.getCanonicalName(), deserializer);
+    }
+
+    /**
+     * Unregister deserialization logic for a given class.
+     *
+     * @param clazz The {@link Class} instance
+     * @see #unregisterDeserialization(String)
+     */
+    public void unregisterDeserialization(@Nonnull final Class<?> clazz) {
+        deserializerMap.remove(clazz.getCanonicalName());
+    }
+
+    /**
+     * Unregister deserialization logic for a given class.
+     *
+     * @param clazz The canonical name of the class instance {@link Class#getCanonicalName()}
+     * @see #unregisterDeserialization(Class)
+     */
+    public void unregisterDeserialization(@Nonnull final String clazz) {
+        deserializerMap.remove(clazz);
+    }
+
+    /**
+     * Attempt to deserialize a given JSON string back to it's object counterpart.
+     *
+     * @param expectedClass The class instance of the expected type
+     * @param json          The JSON string to deserialize
+     * @param <T>           The expected type of the deserialized object
+     * @return Return an optional populated with the deserialized object, else it is empty.
+     *
+     * @see #deserialize(Class, JsonElement)
+     */
+    @Nonnull
+    public <T> Optional<T> deserialize(@Nonnull final Class<T> expectedClass, @Nonnull final String json) {
+        return deserialize(expectedClass, parser.parse(json));
+    }
+
+    /**
+     * Attempt to deserialize a given JSON string back to it's object counterpart.
+     *
+     * @param expectedClass The class instance of the expected type
+     * @param json          The {@link JsonElement} to deserialize
+     * @param <T>           The expected type of the deserialized object
+     * @return Return an optional populated with the deserialized object, else it is empty.
+     */
+    public <T> Optional<T> deserialize(@Nonnull final Class<T> expectedClass, @Nonnull final JsonElement json) {
+        // Try to get the mapped deserialization logic.
+        final Function<JsonElement, ?> function = deserializerMap.get(expectedClass.getCanonicalName());
+        if (function == null) {
+            if (expectedClass.isEnum()) { // Only try generic serialization if explicit deserialization logic is not present.
+                final Object object = deserializeEnum(json);
+                if (object != null) { // Return the successfully reconstructed object instance.
+                    return Optional.of(expectedClass.cast(object));
+                }
+            }
+            return Optional.empty();
+        }
+        return Optional.ofNullable(expectedClass.cast(function.apply(json)));
+    }
```

Comment:
Lots of formatting sins in here.
Comments should never go on the same line as code.
```suggestion
    public <T> Optional<T> deserialize(@Nonnull Class<T> expectedClass, @Nonnull JsonElement json) {
        // Try to get the mapped deserialization logic.
        Function<JsonElement, ?> function = deserializerMap.get(expectedClass.getCanonicalName());

        if (function == null) {
            // Only try generic serialization if explicit deserialization logic is not present.
            if (expectedClass.isEnum()) {
                Object object = deserializeEnum(json);

                // Return the successfully reconstructed object instance.
                if (object != null) { 
                    return Optional.of(expectedClass.cast(object));
                }
            }
            return Optional.empty();
        }

        return Optional.ofNullable(expectedClass.cast(function.apply(json)));
    }
```